geopandas/sindex.py

```diff
@@ -522,7 +547,16 @@ def query(self, geometry, predicate=None, sort=False):
                 input_geometry_index = []
 
                 for i, geo in enumerate(geometry):
-                    res = self.query(geo, predicate=predicate, sort=sort)
+                    if hasattr(distance, "__iter__"):
+                        if len(distance) != len(geometry):
```

Comment:
This doesn't seem quite right: you may have multiple geometries and a single distance value in an array instead of as a scalar; it would still be broadcastable in that case but fail here.

In case it helps, in shapely it always [makes geometry into an array](https://github.com/shapely/shapely/blob/main/shapely/strtree.py#L236-L239) (but keeps track of whether or not it is scalar for deciding the shape of the return values), and distance if provided is [always broadcast to match that](https://github.com/shapely/shapely/blob/main/shapely/strtree.py#L260-L263).

It looks like you should always make the geometry into an array here as well, because it fails if provided a list (pre-existing issue separate from this PR):

```Python
import geopandas
from shapely.geometry import Point, box
from geopandas.sindex import RTreeIndex

s = geopandas.GeoSeries(geopandas.points_from_xy(range(10), range(10)))
sindex = RTreeIndex(s.geometry.values)

sindex.query([box(1, 1, 3, 3)], predicate="dwithin", distance=2)
```

fails with `TypeError`:
```Python
Traceback (most recent call last):
...
    raise TypeError(
TypeError: Got `geometry` of type `<class 'list'>`, `geometry` must be a shapely geometry.
```

whereas it works fine if passed an array.

```Python
sindex.query(np.array([box(1, 1, 3, 3)]), predicate="dwithin", distance=2)
array([[0, 0, 0, 0, 0],
       [0, 1, 2, 3, 4]])
```

Only considering this fix for a possible fix here because you are now trying to broadcast distance to match it, which is easier when geometry is an array.

Reply:
Ah I see, thanks. I've now adapted the code from shapely, with some differences: making geometry into an array only happens for Rtree geometries when predicate = "dwithin",